packages/xo-server-sdn-controller/src/index.js

```diff
@@ -166,41 +166,45 @@ class SDNController extends EventEmitter {
     )
 
     // FIXME: we should monitor when xapis are added/removed
-    forOwn(this._xo.getAllXapis(), async xapi => {
-      await xapi.objectsFetched
+    await Promise.all(
+      map(this._xo.getAllXapis(), async xapi => {
+        await xapi.objectsFetched
+        if (this._setControllerNeeded(xapi)) {
+          return
+        }
 
-      if (this._setControllerNeeded(xapi) === false) {
         this._cleaners.push(await this._manageXapi(xapi))
-
         const hosts = filter(xapi.objects.all, { $type: 'host' })
         await Promise.all(
-          map(hosts, async host => {
+          map(hosts, host => {
             this._createOvsdbClient(host)
```

Comment:
You need to return a promise otherwise there's not point in using `Promise.all`.